spec/models/upload_spec.rb

```diff
@@ -219,256 +239,152 @@
-          it {
-            original_chunks = subject.chunks.all
-            resp = HTTParty.head(
-              "#{subject.storage_provider.storage_url}/#{subject.sub_path}",
-              headers: subject.storage_provider.auth_header
-            )
-            expect(resp.response.code.to_i).to eq(200)
-            subject.chunks.each do |chunk|
-              resp = HTTParty.head(
-                "#{subject.storage_provider.storage_url}/#{chunk.sub_path}",
-                headers: subject.storage_provider.auth_header
-              )
-              expect(resp.response.code.to_i).to eq(200)
-            end
-            purge_time = DateTime.now
+          expect(mocked_storage_provider).to receive(:purge)
+            .with(subject)
+
+          purge_time = DateTime.now
+          expect {
             expect {
               subject.purge_storage
-            }.to change{Chunk.count}.by(-original_chunks.length)
-            subject.reload
-            expect(subject.purged_on).not_to be_nil
-            expect(subject.purged_on).to be >= purge_time
-            expect(subject.chunks.count).to eq(0)
-
-            resp = HTTParty.head(
-              "#{subject.storage_provider.storage_url}/#{subject.sub_path}",
-              headers: subject.storage_provider.auth_header
-            )
-            expect(resp.response.code.to_i).to eq(404)
-            original_chunks.each do |chunk|
-              resp = HTTParty.head(
-                "#{subject.storage_provider.storage_url}/#{chunk.sub_path}",
-                headers: subject.storage_provider.auth_header
-              )
-              expect(resp.response.code.to_i).to eq(404)
-            end
-          }
+            }.to change{Chunk.count}.by(-original_chunks_count)
+          }.not_to raise_error
+
+          subject.reload
+          expect(subject.purged_on).not_to be_nil
+          expect(subject.purged_on).to be >= purge_time
+          expect(subject.chunks.count).to eq(0)
         end
-      end #calls
+      end
     end #purge_storage
-  end #swift methods
+  end #StorageProvider Methods
+
+  describe 'Default StorageProvider' do
+    context 'basic upload' do
+      it 'should be valid when created with a swift_storage_provider' do
+        StorageProvider.delete_all
+        default_storage_provider = FactoryBot.create(:swift_storage_provider, :default)
+        expect(StorageProvider.default).not_to be_nil
+        upload = FactoryBot.create(:upload, storage_provider: StorageProvider.default)
+        expect(upload).to be_valid
+      end
+    end
+
+    context 'completed upload' do
+      it 'should be valid when created with a swift_storage_provider' do
+        StorageProvider.delete_all
+        default_storage_provider = FactoryBot.create(:swift_storage_provider, :default)
+        expect(StorageProvider.default).not_to be_nil
+        upload = FactoryBot.create(:upload, storage_provider: StorageProvider.default)
```

Comment:
The above four lines could be moved to a before block, outside of the current context because it's used in both contexts inside 'Default StorageProvider':
```suggestion
      before(:example) do
        StorageProvider.delete_all
        default_storage_provider = FactoryBot.create(:swift_storage_provider, :default)
        expect(StorageProvider.default).not_to be_nil
      end
      let(:upload) { FactoryBot.create(:upload, storage_provider: StorageProvider.default) }
```
or even:
```suggestion
      before(:example) do
        StorageProvider.delete_all
        expect(default_storage_provider).to be_persisted
        expect(StorageProvider.default).to eq default_storage_provider
      end
      let(:default_storage_provider) { FactoryBot.create(:swift_storage_provider, :default) }
      let(:upload) { FactoryBot.create(:upload, storage_provider: StorageProvider.default) }
```

Reply:
good idea. I improved on it a bit.